date.go

```diff
@@ -5,16 +5,21 @@
 package validation
 
 import (
-	"errors"
 	"time"
 )
 
+var (
+	// ErrDateInvalid is the error that returns in case of an invalid date.
+	ErrDateInvalid = NewError("validation_date_invalid", "must be a valid date")
+	// ErrDateOutOfRange is the error that returns in case of an invalid date.
+	ErrDateOutOfRange = NewError("validation_date_out_of_range", "the data is out of range")
+)
+
 // DateRule is a validation rule that validates date/time string values.
 type DateRule struct {
-	layout       string
-	min, max     time.Time
-	message      string
-	rangeMessage string
+	layout        string
+	min, max      time.Time
+	err, rangeErr Error
```

Comment:
The type should be `error`, and `ErrorObject()` should also use `error` as the parameter type. The reason for this is to support user-defined error types which may allow more advanced i18n features (e.g. plural forms) or other features. 

Note that because of this change, `Error()` method of the rule needs to be implemented differently.

Reply:
Yeah, see my comment on `ErrorObject` please.